### PR TITLE
GMSE: Gradient-Weighted Pressure Loss for High-Gradient Regions

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current loss treats all mesh nodes equally regardless of local field gradient magnitude. In CFD, the physically critical (and hardest-to-predict) regions are exactly where ∇p is large: the leading-edge stagnation, suction peak, trailing-edge separation, and the tandem inter-foil slot. Small prediction errors in high-gradient regions have disproportionate physical impact — yet they receive the same gradient signal as easy freestream nodes.

**GMSE (Gradient MSE):** Weight each node's pressure loss by local field gradient magnitude: `loss_GMSE = (||∇p_pred||₂ + ε) * |p_pred - p_true|`. This automatically amplifies gradient signal in high-gradient regions every batch — without requiring manual identification of hard zones.

**Key distinction from hard-node mining:** Hard-node mining weights by ERROR magnitude (1.5x above-median error nodes). GMSE weights by GRADIENT magnitude. Complementary mechanisms: a low-error node in a high-gradient region still deserves more signal because small errors there have large physical consequences. UNTRIED in 1903 experiments.

**Backed by:** Bhatt et al. "GMSE: Gradient-Weighted MSE for Improved Learning of Complex Physical Fields" (arXiv:2411.17059, 2024) — 5-23% MAE reduction on heat/flow fields vs standard MSE.

**Why it specifically helps p_tan:** The tandem inter-foil slot has strong pressure gradient signatures from fore-foil trailing-edge vortex. GMSE guarantees slot nodes receive 3-5x more gradient signal every batch, not just when they happen to have high error.

## Instructions

Apply gradient weighting to the PRESSURE component of the volume loss only (Ux/Uy less affected by sharp pressure gradients). Compute gradient magnitude via finite differences on mesh edges.

### Step 1: Gradient magnitude via mesh edges

```python
def pressure_gradient_magnitude(pred_p, batch_pos, edge_index, eps=0.1):
    """
    pred_p:     [B*N] predicted pressure (flattened)
    batch_pos:  [B*N, 2] node positions
    edge_index: [2, E] mesh edge connectivity (from batch.edge_index)
    Returns:    [B*N] gradient magnitude at each node
    """
    src, dst = edge_index
    dp = pred_p[dst] - pred_p[src]                  # [E] pressure difference
    dx = batch_pos[dst] - batch_pos[src]             # [E, 2] position difference
    dist = dx.norm(dim=-1).clamp(min=1e-6)           # [E]
    grad_edge = dp.abs() / dist                      # [E] gradient magnitude along edge

    # Scatter-max to nodes: captures the sharpest gradient direction
    grad_node = torch.zeros(pred_p.shape[0], device=pred_p.device)
    grad_node.scatter_reduce_(0, src, grad_edge, reduce='amax', include_self=True)

    # Normalize: weight = (grad_mag + eps), capped to avoid extreme outliers
    grad_weight = (grad_node + eps).clamp(max=grad_node.mean() * 10.0)
    grad_weight = grad_weight / grad_weight.mean()   # normalize: mean=1
    return grad_weight  # [B*N]
```

### Step 2: Apply to pressure volume loss (with stop-gradient)

```python
if args.gmse_pressure_weight > 0.0:
    with torch.no_grad():  # stop-gradient: weights are from current pred, not backpropped
        grad_weight = pressure_gradient_magnitude(
            pred_p_vol.detach().flatten(),
            batch.pos,
            batch.edge_index
        )  # [B*N]
        grad_weight = grad_weight.view_as(pred_p_vol)  # reshape to match vol loss shape

    # Interpolate between standard loss and GMSE:
    # loss = (1 - lambda) * L1(p) + lambda * GMSE(p)
    p_vol_loss_gmse = (pred_p_vol - true_p_vol).abs() * grad_weight
    p_vol_loss = (1 - args.gmse_pressure_weight) * p_vol_loss_standard \
               + args.gmse_pressure_weight * p_vol_loss_gmse
```

**IMPORTANT:** Find exactly where the pressure component of `vol_loss` is computed in `train.py`. Apply GMSE to that component before reduction. Use `torch.no_grad()` for the gradient weight computation (it uses the predicted field which is in the computation graph).

### Step 3: Add the flag

```python
parser.add_argument('--gmse_pressure_weight', type=float, default=0.0,
                    help='GMSE weight: interpolate between L1 (0) and gradient-weighted L1 (1). Start with 0.3.')
```

### Step 4: Run with lambda=0.3 (recommended starting point)

The lambda=0.3 interpolation means 30% of the pressure vol loss is gradient-weighted, 70% is standard. This is conservative — if the run looks good, lambda=0.5 or 0.7 can be tried in a follow-up.

### Sanity check

Before full training, verify on 1 batch:
- Print `grad_weight.min(), grad_weight.max(), grad_weight.mean()`. Expect max 5-15x, mean ~1.0.
- Inspect whether surface/LE nodes have higher weights than freestream nodes.

### Training commands

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent thorfinn --seed 42 \
  --wandb_name "thorfinn/gmse-grad-s42" \
  --wandb_group "gmse-gradient-loss" \
  --gmse_pressure_weight 0.3 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same but CUDA_VISIBLE_DEVICES=1, --seed 73, --wandb_name "thorfinn/gmse-grad-s73"
```

If both seeds converge well and p_tan improves, try a follow-up with `--gmse_pressure_weight 0.5` in the same W&B group.

## Baseline

Current baseline (PR #2290, Re-Stratified Sampling, merged 2026-04-08):

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.74** | < 11.74 |
| p_oodc | 7.65 | < 7.65 |
| **p_tan** | **27.90** | < 27.90 |
| p_re | 6.40 | < 6.40 |

**Reference:** Bhatt et al. "GMSE: Gradient-Weighted MSE for Improved Learning of Complex Physical Fields" (arXiv:2411.17059, 2024)

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2298 because ADVISOR: Closing — GMSE gradient computation on unstructured meshes has proven too complex to implement in the current student session model. No W&B runs were started after multiple pod restarts. The core idea (gradient-weighted loss) remains valid but needs a simpler implementation strategy. Reassigning thorfinn to a different high-impact experiment..
